turbopack/crates/turbo-tasks-fs/src/watcher.rs

```diff
@@ -46,76 +76,91 @@ impl DiskWatcherInternal {
     }
 }
 
-#[derive(Default, Serialize, Deserialize)]
+#[derive(Serialize, Deserialize)]
 pub(crate) struct DiskWatcher {
-    #[serde(skip)]
-    watcher: Mutex<Option<DiskWatcherInternal>>,
-
     /// Array of paths that should not notify invalidations.
     /// `notify` currently doesn't support unwatching subpaths from the root,
     /// so underlying we still watches filesystem event but only skips to
     /// invalidate.
     ignored_subpaths: Vec<PathBuf>,
 
-    /// Keeps track of which directories are currently watched. This is only
-    /// used on OSs that doesn't support recursive watching.
-    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
     #[serde(skip)]
-    watching: dashmap::DashSet<PathBuf>,
+    internal: Mutex<Option<DiskWatcherInternal>>,
+
+    #[serde(skip)]
+    pub(crate) non_recursive_state: Option<NonRecursiveDiskWatcherState>,
```

Comment:
This will switch to recursive mode when restoring from persistent cache. You want to provide a `serde(default = ...)` to initialize it with the right value based on `WATCH_RECURSIVE_MODE`

Reply:
Good catch!